### PR TITLE
Respect accessControl in nodemanagement service, Fix #1177

### LINKED ISSUE #1177: AccessControl Plugin: allowAddNode/Reference no effect
Currently you can set access rights in the Access Control Plugin (https://github.com/open62541/open62541/blob/394df3be8d6a19781927f0790d9c2a9142a2d6a8/include/ua_plugin_access_control.h#L54) but the settings for the following rights are not used:

* allowAddNode
* allowAddReference
* allowDeleteNode
* allowDeleteReference

This is an open issue and can be grabbed by an interested developer.

### PR DESCRIPTION
**Pull request by ccvca**
Fix Issue #1177

The extra functions (restricted*) prevent checking access rights when adding nodes/references by the server.

### CODE REVIEW

File: examples/access_control/client_access_controll.c [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/* This work is licensed under a Creative Commons CCZero 1.0 Universal License.
```
Comment (Pro) [inline examples/access_control/client_access_controll.c]: please rename to `client_access_control.c` (only one L at the end)

File: examples/access_control/server_access_controll.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/* This work is licensed under a Creative Commons CCZero 1.0 Universal License.
```
Comment (Pro) [inline examples/access_control/server_access_controll.c]: please rename to `server_access_control.c` (only one L at the end)

File: examples/access_control/client_access_controll.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * See http://creativecommons.org/publicdomain/zero/1.0/ for more information. */
+
+/**
+ * Using access_control_server
+ */
+
+#include <stdio.h>
+
+#ifdef UA_NO_AMALGAMATION
```
Comment (Pro) [inline examples/access_control/client_access_controll.c]: only use the open62541.h in examples.

Refer to other examle files to see how it is used there.

File: examples/access_control/server_access_controll.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * See http://creativecommons.org/publicdomain/zero/1.0/ for more information. */
+
+#include <signal.h>
+#include <stdio.h>
+
+#ifdef UA_NO_AMALGAMATION
```
Comment (Pro) [inline examples/access_control/server_access_controll.c]: only use the amalgamated header here

**Comment by ccvca** [comment]
The requested changes are forced pushed.

I've implement the changes at the original methods first and noticed that they are called way to often, because when the Server adds a node, the permission-Check is executed.

How can I determine if the node has been added by server or by a client? Does the server always uses the adminSession? If so, it could be integrated in the original methods.

**Comment by ccvca** [comment]
```
/* Do not check access for server */
	if (session != &adminSession &&
		server->config.accessControl.allowDeleteNode &&
		!server->config.accessControl.allowDeleteNode(&session->sessionId, session->sessionHandle,
			item))
	{
....
```
When I integrate the check in the current methods the output of the example looks like this, because addNode also adds a 2 references (forward/backward)?
Is this the intended behaviour? What should happen if userAccess allows adding Nodes but do not allow to add References?

```
Called allowAddNode
Called allowAddReference
Called allowAddReference
Called allowAddReference
Called allowDeleteReference
Called allowDeleteNode
```

**Comment by jpfr** [comment]
Only the `UA_Server_` methods use the adminSession.
For the attributes services, we allow all operations for the adminSession.
This is hard-coded as you suggested: https://github.com/open62541/open62541/blob/master/src/server/ua_services_attribute.c#L22

@Pro I can adjust #1168 after this PR is merged. Still waiting for feedback from Aachen.
In general, we should make the process for first-time contributors as easy as possible. 😄

**Comment by ccvca** [comment]
Ok, so the question is, what is the desired behaviour?
Should UA_Client_addXXXNode result in checks of allowAddNode and allowAddReference (for the forward/backward references) or should it just result in a check of allowAddNode?

The first approach sounds more logical for advanced users,
the second is more straight forward _one Service_ <-> _one permission_

**Comment by jpfr** [comment]
The first approach looks good to me.
Currently, when there is an error during instantiation, the new node is removed...

**Comment by ccvca** [comment]
Ok, the first approach is (force) pushed to this request and the errors of Travis/AppVeyor are fixed.

**Comment by jpfr** [comment]
From what I see, @Pro's comments were taken into account.
The PR looks perfect.

**Comment by jpfr** [comment]
Thanks for the contribution @ccvca 😄

**Pull request merged**